src/Parlot/Fluent/ScopedParser.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System;
+
+namespace Parlot.Fluent
+{
+    /// <summary>
+    /// Returns a new <see cref="Parser{U}" /> converting the input value of 
+    /// type T to the output value of type U using a custom function.
+    /// </summary>
+    /// <typeparam name="T">The input parser type.</typeparam>
+    public sealed class ScopedParser<T> : Parser<T>
+    {
+        private readonly Parser<T> _parser;
+
+        public ScopedParser(Parser<T> parser)
+        {
+            _parser = parser ?? throw new ArgumentNullException(nameof(parser));
+        }
+
+        public override bool Parse(ParseContext context, ref ParseResult<T> result)
+        {
+            context.EnterParser(this);
+
+            return _parser.Parse(new ParseContext(context), ref result);
```

Comment:
Now I understand the goal. However this won't work with contexts that inherit from `ParseContext`, and you could say that the Dictionary is helping there. But my goal was to be able to make the ParserContext a generic type, to make this technique more obvious. A solution could be to add a virtual `Clone` method and use it instead of the valued constructor. Then I would suggest to keep the dictionary out of this PR, and add it to you own `ParserContext` implementation. Once the strongly typed context support is added we could also provide one that has a dictionary.